### PR TITLE
ci: ruff format, isort, and lint

### PR DESCRIPTION
**Pull request by kiyoon**
Instead of black/flake8/pylint, use ruff format, isort and lint as mentioned briefly in #406.

Changes:

1. precommit.sh to use ruff
2. github actions CI to check if ruff format and isort are applied. If not, show the diff in the summary.
4. `ruff check --config ruff_essential.toml .` to check essential syntax errors. (E9, F63, F7, F82)
5. Some settings in pyproject.toml, setup.cfg, ...
    a. move pytest config from pytest.ini to pyproject.toml
    b. move coverage config from setup.cfg to pyproject.toml
    c. use ruff format line-length of 79
    d. ruff lint rules
    e. pyright/basedpyright config (standard settings) to develop with a type checker

Build:
1. Correctly set `python_requires` (3.5 -> 3.8)
2. CI uses `uv pip` to install packages instead of `pip` for improved speed

Advantages:

- fast
- less dependencies (only one package compared to black, isort, flake8, pylint and extra dependencies they have)
- generally easier to config (sane defaults and good documentation in one place)

### CODE REVIEW

**Comment by kiyoon** [comment]
Applying ruff format and isort will change many files. Once you approve this PR, I will apply them all so it's easier to see what has actually changed!

So please don't merge just yet.

File: pdb2pqr/quatfit.py
Comment (sobolevnrm) [inline pdb2pqr/quatfit.py:23]: Is `annotations` used in this module?

File: pdb2pqr/residue.py
Comment (sobolevnrm) [inline pdb2pqr/residue.py:7]: Is `annotations` used in this module?

File: pdb2pqr/run.py
Comment (sobolevnrm) [inline pdb2pqr/run.py:3]: Is `annotations` used in this module?

File: pdb2pqr/structures.py
Comment (sobolevnrm) [inline pdb2pqr/structures.py:11]: Is `annotations` used in this module?

File: pdb2pqr/topology.py
Comment (sobolevnrm) [inline pdb2pqr/topology.py:11]: Is `annotations` used in this module?

File: pdb2pqr/utilities.py
Comment (sobolevnrm) [inline pdb2pqr/utilities.py:14]: Is `annotations` used in this module?

File: pytest.ini
Comment (sobolevnrm) [inline pytest.ini:1]: Why was this deleted?

File: setup.cfg
Comment (sobolevnrm) [inline setup.cfg:4]: Why were the `coverage` parts removed?

File: setup.py
Comment (sobolevnrm) [inline setup.py:28]: Is `annotations` used in this module?

File: setup.py
Comment (sobolevnrm) [inline setup.py:38]: Why was version checking removed?

File: tests/common.py
Comment (sobolevnrm) [inline tests/common.py:3]: Is `annotations` used in this module?

File: tests/core_test.py
Comment (sobolevnrm) [inline tests/core_test.py:3]: Is `annotations` used in this module?

File: tests/io_test.py
Comment (sobolevnrm) [inline tests/io_test.py:3]: Is `annotations` used in this module?

File: tests/ligand_results.py
Comment (sobolevnrm) [inline tests/ligand_results.py:3]: Is `annotations` used in this module?

File: tests/ligand_test.py
Comment (sobolevnrm) [inline tests/ligand_test.py:3]: Is `annotations` used in this module?

File: tests/logging_test.py
Comment (sobolevnrm) [inline tests/logging_test.py:3]: Is `annotations` used in this module?

File: tests/propka_test.py
Comment (sobolevnrm) [inline tests/propka_test.py:3]: Is `annotations` used in this module?

File: tests/regression_test.py
Comment (sobolevnrm) [inline tests/regression_test.py:3]: Is `annotations` used in this module?

File: tests/test_version.py
Comment (sobolevnrm) [inline tests/test_version.py:1]: Is `annotations` used in this module?

**Review by sobolevnrm — CHANGES_REQUESTED** [review]

File: .github/workflows/python-package.yml
Comment (kiyoon) [reply .github/workflows/python-package.yml:28]: I changed it to use `uv` instead to install the package, which is much faster than pip so it will reduce the CI time a bit. If you mind, I can revert.

File: .github/workflows/python-package.yml
Comment (kiyoon) [reply .github/workflows/python-package.yml:29]: Because in `setup.py`, they're already defined as the dependency. So installing with `pip install '.[test]'` will already install them. It's a good practice to put dependencies in one place rather than having them in multiple places.

File: .github/workflows/python-package.yml
Comment (kiyoon) [reply .github/workflows/python-package.yml:14]: The original shell in github action is normall `sh` not `bash`. This changes to use bash as default, and thus can use `~/.bash_profile` to activate environment each step. I saw this first at https://github.com/conda-incubator/setup-miniconda

File: docs/source/conf.py
Comment (kiyoon) [reply docs/source/conf.py:13]: Due to this [ruff rule](https://docs.astral.sh/ruff/rules/missing-required-import/) this got automatically inserted.

If you want to target python < 3.10, this is good to put pretty much everywhere so `Optional[str]`, `Union[str, int]` and be written as `str | None`, `str | int` for example.

File: pdb2pqr/__init__.py
Comment (kiyoon) [reply pdb2pqr/__init__.py:14]: The `setup.py` already states the python target version. `pip` won't ever install the package under python 3.8. This is pretty much redundant and not really a good practice to put those restrictions each file.

File: pdb2pqr/ligand/__init__.py
Comment (kiyoon) [reply pdb2pqr/ligand/__init__.py:12]: See above comment about `setup.py`!

File: pytest.ini
Comment (kiyoon) [reply pytest.ini:1]: It was moved to `pyproject.toml` which is more modern standard of putting all configurations.

File: setup.cfg
Comment (kiyoon) [reply setup.cfg:4]: It was moved to `pyproject.toml`

File: setup.py
Comment (kiyoon) [reply setup.py:38]: This is a wrong way of defining minimum version. The info should go to `python_requires` field in `setup()`. If you put it as a conditional, `pip install` will have a wrong behaviour (just ignoring them).

See https://github.com/neovim/pynvim/issues/582#issuecomment-2470708148, for example.

**Comment by kiyoon** [comment]
by the way, I plan to add more type information in the next PR

**Comment by sobolevnrm** [comment]
> by the way, I plan to add more type information in the next PR.
> 
> If you mind having annotations import, we can remove the lint rule

Yes, I do not want to import unused modules.  Thank you!

File: docs/source/conf.py
Comment (sobolevnrm) [reply docs/source/conf.py:13]: We don't actually use `annotations`, so I would prefer not to include it.

File: .github/workflows/python-package.yml
Comment (sobolevnrm) [reply .github/workflows/python-package.yml:28]: Yes, please revert.  I want to minimize the number of dependencies to those that are absolutely essential.

**Comment by kiyoon** [comment]
I removed `from __future__ import annotations` and `uv`. Anything else?

File: pdb2pqr/__init__.py
Comment (sobolevnrm) [inline pdb2pqr/__init__.py:1]: Please remove the period from the end of the title, unless it's required by the formatting guidelines (in which case, we should discuss).  Thanks.

**Review by sobolevnrm — CHANGES_REQUESTED** [review]

File: pdb2pqr/__init__.py
Comment (kiyoon) [reply pdb2pqr/__init__.py:1]: https://docs.astral.sh/ruff/rules/ends-in-period/

It's because of this rule. Apparently the PEP, numpy and google style guides all follow this rule. If you don't like it, we can remove this rule.

Comment (kiyoon) [reply pdb2pqr/__init__.py]: I removed the period and also added D400 as an ignored rule. Do not worry about it, I think it's quite useless too.

Anything else?

**Review by sobolevnrm — APPROVED** [review]

**Comment by sobolevnrm** [comment]
Thank you for all of your work on this!

**Comment by kiyoon** [comment]
Thanks for reviewing! And I've rebased the #410 so it should be ready now as well :)

**Pull request merged**
